### PR TITLE
Handle args correctly in KUBECTL_EXTERNAL_DIFF

### LINKED ISSUE #1178: kubectl diff doesn’t handle quoted options in KUBECTL_EXTERNAL_DIFF correctly
### What happened?

I tried to make a diff ignoring lines matching a regex which includes special characters and thus has to be quoted:
```
KUBECTL_EXTERNAL_DIFF="diff -I\"^  generation: [0-9]\+$\"" kubectl diff -f filename.yaml
```
The matching line was not ignored in the output.

To get to the bottom of this I tried it with `KUBECTL_EXTERNAL_DIFF="diff -Igeneration"` and the matching change was successfully ignored, however when I added quotes like `"diff -I\"generation\""` or `'diff -I"generation"'` it failed to ignore the line.

### What did you expect to happen?

`kubectl diff` handles the quoting correctly so that these are equivalent:
```
KUBECTL_EXTERNAL_DIFF="diff -Igeneration"
KUBECTL_EXTERNAL_DIFF="diff -I\"generation\""
KUBECTL_EXTERNAL_DIFF='diff -I"generation"'
```
It thus allows for special characters in the `KUBECTL_EXTERNAL_DIFF` command options.

### How can we reproduce it (as minimally and precisely as possible)?

Try ignoring a change in the diff by using a quoted string inside `KUBECTL_EXTERNAL_DIFF`.

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.3", GitCommit:"816c97ab8cff8a1c72eccca1026f7820e93e0d25", GitTreeState:"archive", BuildDate:"2022-01-27T18:26:18Z", GoVersion:"go1.17.6", Compiler:"gc", Platform:"linux/amd64"}
```
</details>

### Cloud provider

<details>
Google
</details>

### OS version

<details>

```console
$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux-logo
$ uname -a
Linux hostname 5.16.8-arch1-1 kubernetes/kubernetes#1 SMP PREEMPT Tue, 08 Feb 2022 21:21:08 +0000 x86_64 GNU/Linux
```

</details>

### Install tools

_No response_

### Container runtime (CRI) and and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by eddiezane**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This PR improves how we are passing arguments to the kubectl diff when using an external diff program.

#### Which issue(s) this PR fixes:

Fixes kubernetes/kubectl#1178

#### Special notes for your reviewer:

The `github.com/mattn/go-shellwords` was a dependency before but looks like it was removed after 1.19. I tried a bunch of different ways to handle the parsing of arguments and flags without it but this functionality is surprisingly not exported in cobra or the stdlib.

#### Does this PR introduce a user-facing change?

```release-note
Improve how arguments are passed for KUBECTL_EXTERNAL_DIFF and remove character restrictions
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/diff_test.go [hunk 1]
```diff
@@ -61,31 +62,81 @@ func (f *FakeObject) Live() runtime.Object {
...
+			}
+		}
+		if err == nil && tc.expectedError != "" {
+			t.Errorf("expected error: %v", tc.expectedError)
+		}
+		if bin != "diff" {
+			t.Errorf("unexpected binary: %s", bin)
+		}
+
+		// TODO: need to PR k8s.io/utils with this
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/diff/diff_test.go:132]: I need to get these changes in https://github.com/kubernetes/utils.

/hold

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go [hunk 2]
```diff
@@ -94,8 +94,8 @@ const (
 
 // diffError returns the ExitError if the status code is less than 1,
 // nil otherwise.
-func diffError(err error) exec.ExitError {
-	if err, ok := err.(exec.ExitError); ok && err.ExitStatus() <= 1 {
+func diffError(err error) *exec.ExitError {
+	if err, ok := err.(*exec.ExitError); ok && err.ExitCode() <= 1 {
```
Comment (eddiezane) [inline staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:98]: I think this is safe to do? It's probably more system independent than the previous wrapper.

https://github.com/kubernetes/utils/blob/3a6ce19ff2f91f3dc49243f90f7834a0f1e4aaf3/exec/exec.go#L218-L225

**Comment by liggitt** [comment]
do we have to handle shell-quoted args with spaces here? I'd prefer not to re-add the dependency and keep this restriction in place

**Comment by eddiezane** [comment]
> do we have to handle shell-quoted args with spaces here? I'd prefer not to re-add the dependency and keep this restriction in place

I'd also rather not add the dependency. Other languages have this functionality as part of the stdlib and I'm really surprised Go doesn't.

That said the functionality of external diff is pretty gimped right now. The current restriction is limited to just alphanumeric and - =. I think the issue author has a pretty valid use case of wanting to specify an ignore filter. This change opens the doors up for quite a bit.

As far as go-shellwords goes it's pretty stable and has no deps itself.

**Comment by liggitt** [comment]
hmm... possible alternatives to a shell-parsing dependency that come to mind:
1. put the command they want in a script and put the name of the script as $KUBECTL_EXTERNAL_DIFF
2. have an alternative envvar that has an encoded string we can parse without shell escaping (e.g. `KUBECTL_EXTERNAL_DIFF_JSON='["command","arg with spaces","another arg with spaces"]'`)
3. same as 2 but use the existing envvar and trigger parsing of the value off whether the value starts and ends with `[...]`

1 doesn't seem terrible as a workaround, and 2 or 3 seem better than trying to emulate shell parsing and taking on an additional dependency

**Comment by eddiezane** [comment]
@liggitt how does this look now? I've preserved backwards functionality and added a warning.

/hold

**Review by liggitt — COMMENTED** [review]
the arg handling changes lgtm (with a couple comments). the switch to going directly to os/exec I'm unsure about

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go [hunk 3]
```diff
@@ -185,40 +192,63 @@ func NewCmdDiff(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 // KUBECTL_EXTERNAL_DIFF environment variable will be used a diff
 // program. By default, `diff(1)` will be used.
 type DiffProgram struct {
-	Exec exec.Interface
```
[the remark this answers is not in the document]
Comment (eddiezane) [reply staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:188]: This change was needed to test the command args at https://github.com/kubernetes/kubernetes/pull/108199/files#diff-d26a728b714127f1de3af27904cc297d34b0bb58be6c023ec5510e28fdda1a49R154.

I think this is a better move than wrapping the interface every time we need to add something like https://github.com/kubernetes/utils/pull/238.

**Comment by eddiezane** [comment]
[Flake](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108199/pull-kubernetes-e2e-kind-ipv6/1501596112572649472) is https://github.com/kubernetes/kubernetes/issues/106904

**Comment by eddiezane** [comment]
[Flake](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/108199/pull-kubernetes-e2e-kind/1501614140739817472) is https://github.com/kubernetes/kubernetes/issues/107374

File: staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:188]: for interfaces used to spawn processes/network connections, an interface is better than embedding the concrete implementation

> This change was needed to test the command args at https://github.com/kubernetes/kubernetes/pull/108199/files#diff-d26a728b714127f1de3af27904cc297d34b0bb58be6c023ec5510e28fdda1a49R154.

That's the beauty of an interface, we don't need to change to use os/exec directly. In the unit test, wrap the Exec with an implementation that records the command and args passed in:

```diff
+                       recordingExec := &recordingExec{Interface: exec.New()}
                        diff := DiffProgram{
+                               Exec:      recordingExec,
                                IOStreams: streams,
                        }
 
@@ -151,7 +154,7 @@ func TestExternalDiff(t *testing.T) {
                                t.Fatalf("expected: %q got: %q", tc.expectedWarning, w)
                        }
                        if cmd != nil {
-                               cmdArgs := cmd.Args
+                               cmdArgs := recordingExec.Args
                                // The last two args are /dev/zero
                                args := cmdArgs[:len(cmdArgs)-2]
                                if !reflect.DeepEqual(tc.expectedArgs, args) {
@@ -162,6 +165,16 @@ func TestExternalDiff(t *testing.T) {
        }
 }
 
+type recordingExec struct {
+       Args []string
+       exec.Interface
+}
+
+func (r *recordingExec) Command(cmd string, args ...string) exec.Cmd {
+       r.Args = append([]string{cmd}, args...)
+       return r.Interface.Command(cmd, args...)
+}
+
```

you can then monitor (and tweak, if you wanted) any calls / inputs / outputs

**Comment by liggitt** [comment]
unassigning, feel free to reassign when ready for review

/unassign

**Pull request closed without merging**